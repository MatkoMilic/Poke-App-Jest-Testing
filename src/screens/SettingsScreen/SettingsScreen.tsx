import React from 'react';
import {Text, TouchableOpacity, Switch} from 'react-native';
import {Button, useTheme} from 'react-native-paper';
import {Header, ScreenContainer, ThemeContext} from '../../components';
import {removeActiveUser} from '../../constants';
import {
  IMainNavScreenProps,
  MainNavigatorScreens,
  NavigatorNames,
  OnboardingNavigatorScreens,
} from '../../types';
import styles from './styles';

interface SettingsScreenProps extends IMainNavScreenProps {}

const SettingsScreen: React.FC<SettingsScreenProps> = ({navigation}) => {
  const goToProfile = () => {
    navigation.navigate(MainNavigatorScreens.PROFILE_SCREEN);
  };
  const goToPokeList = () => {
    navigation.navigate(MainNavigatorScreens.POKELIST_SCREEN);
  };
  const theme = useTheme();
  const {toggleTheme, isThemeDark} = React.useContext(ThemeContext);

  const logoutUser = async () => {
    removeActiveUser();
    navigation.replace(NavigatorNames.ONBOARDING_NAVIGATOR, {
      screen: OnboardingNavigatorScreens.LOGIN_SCREEN,
    });
  };

  return (
    <ScreenContainer>
      <Header
        headerTitle="Poke Settings"
        leftIcon="home-account"
        rightIcon="clipboard-list"
        headerSubtitle="2front"
        leftOnPress={goToProfile}
        rightOnPress={goToPokeList}
      />
      <TouchableOpacity>
        <Switch
          testID="testSwitch"
          style={styles.switchElement}
          value={isThemeDark}
          onValueChange={toggleTheme}
        />
      </TouchableOpacity>
      <Text testID={'isThemeDarkText'} style={styles.themeText_TESTING}>
        {`Theme at the moment is: ${isThemeDark.toString()}`}
      </Text>
      <Text>Welcome to settings</Text>
      <Button onPress={logoutUser}>LOG OUT</Button>
    </ScreenContainer>
  );
};

export default SettingsScreen;
