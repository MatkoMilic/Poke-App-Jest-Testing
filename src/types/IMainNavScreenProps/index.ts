export type {
  IMainNavScreenProps,
  MainNavigationType,
} from './IMainNavScreenProps';
