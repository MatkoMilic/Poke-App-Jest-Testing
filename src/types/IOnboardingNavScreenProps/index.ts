export type {
  IOnboardingNavScreenProps,
  OnboardingNavigationType,
} from './IOnboardingNavScreenProps';
