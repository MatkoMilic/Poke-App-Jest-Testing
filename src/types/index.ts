export type {
  IMainNavScreenProps,
  MainNavigationType,
} from './IMainNavScreenProps';
export type {
  IOnboardingNavScreenProps,
  OnboardingNavigationType,
} from './IOnboardingNavScreenProps';
export type {
  MainNavigatorKeys,
  NavigatorNameKeys,
  OnboardingNavigatorParamList,
  RootNavigatorKeys,
  RootNavigatorParamList,
  OnboardingNavigatorKeys,
  MainNavigatorParamList,
  RootNavigatorType,
} from './navigatorTypes';
export {
  NavigatorNames,
  OnboardingNavigatorScreens,
  MainNavigatorScreens,
} from './navigatorTypes';
export type {IUserValues} from './IUserValues/IUserValues';
export {initialUserLoginValues} from './IUserValues/IUserValues';
