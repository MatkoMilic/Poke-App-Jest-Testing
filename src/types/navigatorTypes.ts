import {NavigatorScreenParams} from '@react-navigation/native';

export enum OnboardingNavigatorScreens {
  LOADING_SCREEN = 'LoadingScreen',
  LOGIN_SCREEN = 'LoginScreen',
}
export type OnboardingNavigatorParamList = {
  LOADING_SCREEN: undefined;
  LOGIN_SCREEN: undefined;
};
export type OnboardingNavigatorKeys = keyof OnboardingNavigatorScreens;
export enum MainNavigatorScreens {
  PROFILE_SCREEN = 'ProfileScreen',
  POKELIST_SCREEN = 'PokeListScreen',
  SETTINGS_SCREEN = 'SettingsScreen',
}
export type MainNavigatorParamList = {
  PROFILE_SCREEN: undefined;
  POKELIST_SCREEN: undefined;
  SETTINGS_SCREEN: undefined;
};
export type MainNavigatorKeys = keyof MainNavigatorScreens;
export enum NavigatorNames {
  MainNavigator = 'MainNavigator',
  OnboardingNavigator = 'OnboardingNavigator',
}
export type NavigatorNameKeys = keyof NavigatorNames;
export type RootNavigatorType = {
  MainNavigator: NavigatorScreenParams<MainNavigatorParamList>;
  OnboardingNavigator: NavigatorScreenParams<OnboardingNavigatorParamList>;
};
export type RootNavigatorParamList = {
  MainNavigator: undefined;
  OnboardingNavigator: undefined;
};
export type RootNavigatorKeys = keyof RootNavigatorType;
