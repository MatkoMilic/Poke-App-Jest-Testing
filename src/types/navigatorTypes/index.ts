export type {
  MainNavigatorKeys,
  OnboardingNavigatorKeys,
  MainNavigatorParamList,
  RootNavigatorParamList,
  RootNavigatorType,
  RootNavigatorKeys,
  NavigatorNameKeys,
  OnboardingNavigatorParamList,
} from './navigatorTypes';
export {
  NavigatorNames,
  OnboardingNavigatorScreens,
  MainNavigatorScreens,
} from './navigatorTypes';
