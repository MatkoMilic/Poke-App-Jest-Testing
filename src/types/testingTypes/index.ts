export type {ReactWrapperType, RenderApiType} from './testingTypes';
