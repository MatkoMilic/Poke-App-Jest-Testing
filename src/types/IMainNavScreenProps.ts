import {CompositeNavigationProp} from '@react-navigation/native';
import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import {RootNavigatorType, MainNavigatorParamList} from './navigatorTypes';

export type MainNavigationType = CompositeNavigationProp<
  NativeStackNavigationProp<MainNavigatorParamList>,
  NativeStackNavigationProp<RootNavigatorType>
>;
export interface IMainNavScreenProps {
  navigation: MainNavigationType;
}
