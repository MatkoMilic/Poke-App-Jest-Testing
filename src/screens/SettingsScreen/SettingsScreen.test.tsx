import 'jsdom-global/register';
import React from 'react';
import {Switch, Text} from 'react-native';
import {fireEvent, render} from '@testing-library/react-native';
import AsyncStorage from '@react-native-async-storage/async-storage';
import {configure, mount} from 'enzyme';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import {MainNavigationType} from '../../types';
import SettingsScreen from './SettingsScreen';
import {
  Header,
  ScreenContainer,
  ThemeContext,
  ThemeProvider,
} from '../../components';
configure({adapter: new Adapter()});

describe('SettingsScreen change theme', () => {
  let navigation: MainNavigationType;
  test('renders correctly', () => {
    const tree = renderer
      .create(<SettingsScreen navigation={navigation} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('does SettingsScreen contain header', () => {
    const mountSettingsScreen = mount(
      <SettingsScreen navigation={navigation} />,
    );
    expect(mountSettingsScreen.containsMatchingElement(<Header />)).toEqual(
      true,
    );
  });
  test('does SettingsScreen contain header', () => {
    const mountSettingsScreen = mount(
      <SettingsScreen navigation={navigation} />,
    );
    expect(mountSettingsScreen).toBeDefined();
  });
  it('first renders with dark theme off (false) because its default', () => {
    const {getByText} = render(
      <ThemeContext.Consumer>
        {(value) => (
          <SettingsScreen navigation={navigation}>
            <Text>{value.isThemeDark}</Text>
          </SettingsScreen>
        )}
      </ThemeContext.Consumer>,
    );
    expect(getByText('false')).toBeTruthy();
  });
  it('test change isThemeDark to true and onValueChange firing', () => {
    const {getAllByText, getByTestId} = render(
      <ThemeProvider>
        <ThemeContext.Consumer>
          {(value) => (
            <ScreenContainer>
              <Text>{value.isThemeDark.toString()}</Text>
              <Switch
                testID={'testSwitch'}
                value={value.isThemeDark}
                onValueChange={value.toggleTheme}
              />
            </ScreenContainer>
          )}
        </ThemeContext.Consumer>
      </ThemeProvider>,
    );
    const themeSwitch = getByTestId('testSwitch');
    fireEvent.press(themeSwitch, 'onValueChange');
    expect(getAllByText('true')[0]).toBeTruthy();
  });
  it('test does saved theme in async storage match the right theme', async () => {
    const mountSettingsScreen = mount(
      <ThemeProvider>
        <ThemeContext.Consumer>
          {(value) => (
            <ScreenContainer>
              <Text>{value.isThemeDark.toString()}</Text>
              <Switch
                testID={'testSwitch'}
                value={value.isThemeDark}
                onValueChange={value.toggleTheme}
              />
            </ScreenContainer>
          )}
        </ThemeContext.Consumer>
      </ThemeProvider>,
    );
    const valueText = mountSettingsScreen.find('Text').first().text();
    await AsyncStorage.setItem('theme', valueText);
    let theme = await AsyncStorage.getItem('theme');
    expect(theme).toBe(valueText);
  });
  it('does it unmount', () => {
    const mountSettingsScreen = mount(
      <ThemeProvider>
        <ThemeContext.Consumer>
          {(value) => (
            <ScreenContainer>
              <Text>{value.isThemeDark.toString()}</Text>
              <Switch
                testID={'testSwitch'}
                value={value.isThemeDark}
                onValueChange={value.toggleTheme}
              />
            </ScreenContainer>
          )}
        </ThemeContext.Consumer>
      </ThemeProvider>,
    );
    mountSettingsScreen.unmount();
    expect(mountSettingsScreen.exists()).toBeFalsy();
  });
});
