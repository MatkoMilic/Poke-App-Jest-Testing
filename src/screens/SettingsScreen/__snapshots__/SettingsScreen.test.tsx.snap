// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsScreen change theme settings screen renders correctly 1`] = `
<RNCSafeAreaView
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Array [
        Object {
          "backgroundColor": "#add8e6",
          "elevation": 4,
          "zIndex": 0,
        },
        Object {
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 3,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 4,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#add8e6",
          "elevation": 0,
          "flexDirection": "row",
          "height": 56,
          "marginTop": 0,
          "paddingHorizontal": 4,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": undefined,
          }
        }
        accessible={true}
        focusable={true}
        hitSlop={
          Object {
            "bottom": 6,
            "left": 6,
            "right": 6,
            "top": 6,
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "margin": 6,
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 22.5,
                "height": 45,
                "width": 45,
              },
              undefined,
              undefined,
            ],
          ]
        }
      >
        <View>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                },
                Object {
                  "height": 30,
                  "width": 30,
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "bottom": 0,
                  "left": 0,
                  "opacity": 1,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "transform": Array [
                    Object {
                      "rotate": "0deg",
                    },
                  ],
                }
              }
            >
              <Text
                accessibilityElementsHidden={true}
                allowFontScaling={false}
                importantForAccessibility="no-hide-descendants"
                pointerEvents="none"
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontSize": 30,
                    },
                    Array [
                      Object {
                        "lineHeight": 30,
                        "transform": Array [
                          Object {
                            "scaleX": 1,
                          },
                        ],
                      },
                      Object {
                        "backgroundColor": "transparent",
                      },
                    ],
                    Object {
                      "fontFamily": "Material Design Icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                󰠦
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "flex": 1,
              "paddingHorizontal": 12,
            },
            Array [
              Object {
                "marginLeft": 8,
              },
              Object {
                "alignItems": "center",
              },
              Object {
                "alignItems": "center",
              },
            ],
          ]
        }
      >
        <Text
          accessibilityRole="header"
          accessibilityTraits="header"
          accessible={true}
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "fontSize": 17,
                },
                undefined,
              ],
            ]
          }
        >
          Poke Settings
        </Text>
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Array [
                Object {
                  "fontSize": 11,
                },
                Object {
                  "color": "rgba(0, 0, 0, 0.7)",
                },
                undefined,
              ],
            ]
          }
        >
          2front
        </Text>
      </View>
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": undefined,
          }
        }
        accessible={true}
        focusable={true}
        hitSlop={
          Object {
            "bottom": 6,
            "left": 6,
            "right": 6,
            "top": 6,
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "margin": 6,
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 22.5,
                "height": 45,
                "width": 45,
              },
              undefined,
              undefined,
            ],
          ]
        }
      >
        <View>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                },
                Object {
                  "height": 30,
                  "width": 30,
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "bottom": 0,
                  "left": 0,
                  "opacity": 1,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "transform": Array [
                    Object {
                      "rotate": "0deg",
                    },
                  ],
                }
              }
            >
              <Text
                accessibilityElementsHidden={true}
                allowFontScaling={false}
                importantForAccessibility="no-hide-descendants"
                pointerEvents="none"
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontSize": 30,
                    },
                    Array [
                      Object {
                        "lineHeight": 30,
                        "transform": Array [
                          Object {
                            "scaleX": 1,
                          },
                        ],
                      },
                      Object {
                        "backgroundColor": "transparent",
                      },
                    ],
                    Object {
                      "fontFamily": "Material Design Icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                󱃔
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="testSwitch"
    >
      <Text>
        false
      </Text>
    </View>
  </View>
  <Text
    style={
      Object {
        "display": "none",
      }
    }
    testID="isThemeDarkText"
  >
    Theme at the moment is: false
  </Text>
  <Text>
    Welcome to settings
  </Text>
  <View
    style={
      Object {
        "backgroundColor": "transparent",
        "borderColor": "transparent",
        "borderRadius": 4,
        "borderStyle": "solid",
        "borderWidth": 0,
        "elevation": 0,
        "minWidth": 64,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 4,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <Text
          numberOfLines={1}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Array [
                Object {
                  "letterSpacing": 1,
                  "marginHorizontal": 16,
                  "marginVertical": 9,
                  "textAlign": "center",
                },
                undefined,
                Object {
                  "textTransform": "uppercase",
                },
                Object {
                  "color": "#6200ee",
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                Object {
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                undefined,
              ],
            ]
          }
        >
          LOG OUT
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
